Feature/add user login

sashakurta opened the pull request:
- Add auth for user by spring secure interface implementation
POST signIn
POST signUp
- Add roles for users
- Refactor deck (from mock user to user model) as well as tests for deck
- Add new units for repository for cards and deck
- Units and la integration test to UserService and SecurityService

ajax-kurta-o committed:
Add CRUD for cards

ajax-kurta-o committed:
Add trigger

ajax-kurta-o committed:
Add trigger

ajax-kurta-o committed:
remove useless unit test for service

ajax-kurta-o committed:
add step to run unit tests and plugin to stdout report to console

ajax-kurta-o committed:
Remove base controller, rename base url, remove html static, refactor pipeline

ajax-kurta-o committed:
Add DTO objects and dto to model map
Add mongodb integration and change repository to mongo
Change docker compose file to add volume
Refactor tests
Change get to get by collection

ajax-kurta-o committed:
fix linter

ajax-kovtunov-o committed:
Fix?

ajax-kurta-o committed:
Remove configuration, fix application.properties and fix 1 test

ajax-kurta-o committed:
Refactor tests, refactor mappings and remove get

ajax-kurta-o committed:
Add user login/sign in with roles functional

ajax-kurta-o committed:
Merge branch 'refs/heads/master' into feature/add-user-login

# Conflicts:
#	.github/workflows/merge_pipeline.yml
#	build.gradle.kts
#	src/main/kotlin/io/github/anki/anki/controller/CardsController.kt
#	src/main/kotlin/io/github/anki/anki/controller/ControllerExceptionHandler.kt
#	src/main/kotlin/io/github/anki/anki/controller/dto/CardDto.kt
#	src/main/kotlin/io/github/anki/anki/controller/dto/mapper/CardDtoMapper.kt
#	src/main/kotlin/io/github/anki/anki/repository/mongodb/CardRepository.kt
#	src/main/kotlin/io/github/anki/anki/repository/mongodb/document/MongoCard.kt
#	src/main/kotlin/io/github/anki/anki/service/CardsService.kt
#	src/main/kotlin/io/github/anki/anki/service/model/Card.kt
#	src/main/kotlin/io/github/anki/anki/service/model/mapper/CardMapper.kt
#	src/main/resources/application.properties
#	src/test/kotlin/io/github/anki/anki/controller/CardsControllerTest.kt
#	src/test/kotlin/io/github/anki/anki/repository/CardRepositoryTest.kt
#	src/test/kotlin/io/github/anki/anki/service/model/mapper/CardMapperKtTest.kt

ajax-kurta-o committed:
Update and refactor dtos

ajax-kurta-o committed:
Revert "Update and refactor dtos"

This reverts commit f63c2cfad95906099e8766bdd93f80df7ff51add.

ajax-kurta-o committed:
Merge branch 'refs/heads/master' into feature/add-user-login

# Conflicts:
#	src/main/kotlin/io/github/anki/anki/repository/mongodb/document/MongoCard.kt

ajax-kurta-o committed:
Update and refactor dtos

ajax-kurta-o committed:
Remove test controller for roles

ajax-kurta-o committed:
Fix tests and detekt

ajax-kurta-o committed:
Fix tests and detekt

ajax-kurta-o committed:
Remove !! not safe null checks

ajax-kurta-o committed:
Fix detekt

Cashtan54 commented on src/main/kotlin/io/github/anki/anki/service/UserService.kt:
```diff
@@ -0,0 +1,58 @@
+@Service
+class UserService : UserDetailsService {
+    @Autowired
+    var userRepository: UserRepository? = null
```
move to constructor

sashakurta replied:
+

sashakurta replied:
+

Cashtan54 commented on src/test/kotlin/io/github/anki/anki/controller/dto/mapper/CardDtoMapperTest.kt:
```diff
@@ -88,7 +88,7 @@ class CardDtoMapperTest {
 
             violations.first().propertyPath.toString() shouldBe "key"
 
-            violations.first().message shouldBe "must not be blank"
```
potuzhno

Cashtan54 commented on src/main/kotlin/io/github/anki/anki/controller/AuthController.kt:
```diff
@@ -0,0 +1,69 @@
+
+    @PostMapping(SIGN_UP)
+    @ResponseStatus(HttpStatus.CREATED)
+    fun registerUser(@RequestBody signUpRequestDto: @Valid SignUpRequestDto?): MessageResponseDto {
```
no need to set SignUpRequestDto as nullable and use !!

sashakurta replied:
+

Cashtan54 commented on src/main/kotlin/io/github/anki/anki/controller/dto/auth/MessageResponseDto.kt:
```diff
@@ -0,0 +1,3 @@
+package io.github.anki.anki.controller.dto.auth
+
+class MessageResponseDto(var message: String)
```
what is MessageResponseDto? name means nothing

sashakurta replied:
change to UserCreatedMessageResponseDto

Cashtan54 commented on src/main/kotlin/io/github/anki/anki/repository/mongodb/RoleRepository.kt:
```diff
@@ -0,0 +1,10 @@
+import org.springframework.data.mongodb.repository.MongoRepository
+import java.util.Optional
+
+interface RoleRepository : MongoRepository<MongoRole?, String?> {
```
lets rewrite to self-made method

sashakurta replied:
i removed because no needed and had no usages

Cashtan54 commented on src/main/kotlin/io/github/anki/anki/repository/mongodb/document/MongoDeck.kt:
```diff
@@ -13,18 +13,18 @@ import java.time.Instant
     @Id
     @Field(MongoDocument.ID)
-    override val id: ObjectId? = null,
+    override var id: ObjectId? = null,
```
why do you want to change id after creating?

sashakurta replied:
it was for tests but i removed

Cashtan54 commented on src/main/kotlin/io/github/anki/anki/repository/mongodb/document/MongoRole.kt:
```diff
@@ -0,0 +1,21 @@
+data class MongoRole(
+    @Id
+    @Field(MongoDocument.ID)
+    var id: ObjectId? = null,
```
should be val

sashakurta replied:
+

Cashtan54 commented on src/main/kotlin/io/github/anki/anki/repository/mongodb/document/MongoUser.kt:
```diff
@@ -0,0 +1,57 @@
+    @Id
+    @Field(MongoDocument.ID)
+    override var id: ObjectId? = null,
+    @NotBlank
```
do you call jakarta validation on this fields?
And why do you want to validate data on mongo level, i guest validate in controller is enough

sashakurta replied:
+

Cashtan54 commented on src/main/kotlin/io/github/anki/anki/service/UserService.kt:
```diff
@@ -0,0 +1,58 @@
+    @Autowired
+    var userRepository: UserRepository? = null
+
+    @Transactional
```
do you need Transactional here? you have only 1 mongo request here

sashakurta replied:
+

Cashtan54 commented on src/main/kotlin/io/github/anki/anki/service/UserService.kt:
```diff
@@ -0,0 +1,58 @@
+    @Throws(UsernameNotFoundException::class)
+    override fun loadUserByUsername(userName: String): UserDetails {
+        val user: MongoUser =
+            userRepository?.findByUserName(userName)
```
delete ? after moving userRepository to constructor

sashakurta replied:
+

Cashtan54 commented on src/main/kotlin/io/github/anki/anki/service/UserService.kt:
```diff
@@ -0,0 +1,58 @@
+    @Transactional
+    @Throws(UsernameNotFoundException::class)
+    override fun loadUserByUsername(userName: String): UserDetails {
+        val user: MongoUser =
```
rename to mongoUser for make it easier to read

sashakurta replied:
+

Cashtan54 commented on src/main/kotlin/io/github/anki/anki/service/UserService.kt:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    fun signIn(user: User): User {
+        if (!userRepository?.existsByUserName(user.userName)!!) {
```
if (userRepository.existsByUserName(user.userName)) {
    return user
}
throw UserDoesNotExistException.fromUserName(user.userName)

sashakurta replied:
тоді вже (userRepository.existsByUserName(user.userName)) все одно з !! 
UPD побачив комент вище

Cashtan54 commented on src/main/kotlin/io/github/anki/anki/service/UserService.kt:
```diff
@@ -0,0 +1,58 @@
+                throw ex
+            }
+        }
+        return null
```
its better to never return null explicitly
this method should fast fall if we cant create user

sashakurta replied:
+

ajax-kovtunov-o commented on config/detekt.yml:
```diff
@@ -126,7 +126,7 @@ complexity:
     active: true
     functionThreshold: 6
-    constructorThreshold: 7
+    constructorThreshold: 10
```
Might we create code style rules to prevent inconsistent code to be applied, instead of adopting the rules when inconsistent code written?

sashakurta replied:
changed to 8, but i if i need more values in constructor, how can i improve it to make less args in this case?

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/controller/AuthController.kt:
```diff
@@ -0,0 +1,69 @@
+
+    companion object {
+        private val LOG: Logger = LoggerFactory.getLogger(AuthController::class.java)
+        const val BASE_URL = "/api/auth/v1"
```
Moving routes to the constants might kind of good coding practice, but from my view it isn't mandatory. The current problem is: if u used constants for one controller, maybe u need to adopt others or keep things as it was.

sashakurta replied:
i do this for all current controllers and for tests for changed contollers)

ajax-kurta-o committed:
Fix detekt

ajax-kurta-o committed:
Fix must not be blank error

ajax-kurta-o committed:
Fix must not be blank error

ajax-kurta-o committed:
Fix encoder initialization for AuthDtoMapperTest

ajax-kurta-o committed:
Fix encoder initialization for AuthDtoMapperTest

ajax-kurta-o committed:
Fix encoder initialization for AuthDtoMapperTest

ajax-kurta-o committed:
Fix pr comments

ajax-kurta-o committed:
Fix controllers and add test for auth to ControllerExceptionHandlerTest

ajax-kurta-o committed:
Fix controllers and add test for auth to ControllerExceptionHandlerTest

ajax-kurta-o committed:
Move Repository to constructor

ajax-kurta-o committed:
Remove nullable

ajax-kurta-o committed:
Change detekt constructor value

ajax-kurta-o committed:
Fix pr comments

ajax-kurta-o committed:
Fix pr comments

ajax-kurta-o committed:
Remove logger

ajax-kurta-o committed:
Remove logger

ajax-kurta-o committed:
Remove logger

ajax-kurta-o committed:
Rollout logger

ajax-kurta-o committed:
Rollout logger

ajax-kurta-o committed:
Change dto attributes to val

ajax-kurta-o committed:
Change dto attributes to val

ajax-kurta-o committed:
Remove SecurityServiceTest.kt

ajax-kurta-o committed:
Fix should sign up User always

ajax-kurta-o committed:
Change to val

ajax-kurta-o committed:
Rise deltaCoverage

ajax-kurta-o committed:
Rise deltaCoverage

ajax-kurta-o committed:
Rise deltaCoverage

ajax-kurta-o committed:
Rise deltaCoverage

ajax-kurta-o committed:
Fix should return 200 if user with email exist but and doesn't create one

ajax-kovtunov-o commented on config/detekt.yml:
```diff
@@ -126,7 +126,7 @@ complexity:
     active: true
     functionThreshold: 6
-    constructorThreshold: 7
+    constructorThreshold: 8
```
🌝

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/controller/AuthController.kt:
```diff
@@ -0,0 +1,69 @@
+    @ResponseStatus(HttpStatus.OK)
+    fun authenticateUser(@RequestBody signInRequestDto: @Valid SignInRequestDto?): JwtResponseDto {
+        LOG.info(
+            "IN: ${AuthController::class.java.name}: ${BASE_URL}${SIGN_IN} with userName${signInRequestDto!!.userName}",
```
don't use !!

ajax-kovtunov-o replied:
It's no any sense to log request path, if u need u can enable it on spring level. U need to log the operations.

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/controller/dto/auth/JwtResponseDto.kt:
```diff
@@ -0,0 +1,13 @@
+    val email: String?,
+    val roles: Set<String>,
+) {
+    val tokenType: TokenType = TokenType.BEARER
```
token type is unused

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/service/secure/SecurityService.kt:
```diff
@@ -0,0 +1,28 @@
+    var jwtUtils: JwtUtils,
+) {
+
+    fun authUser(user: SignInRequestDto): Authentication {
```
don't use `SignInRequestDto` in service layer, pls map to service layer model instead

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/service/secure/SecurityService.kt:
```diff
@@ -0,0 +1,28 @@
+                UsernamePasswordAuthenticationToken(user.userName, user.password),
+            )
+        SecurityContextHolder.getContext().setAuthentication(authentication)
+        return authentication
```
add log that user is authenticated

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/service/secure/SecurityService.kt:
```diff
@@ -0,0 +1,28 @@
+            authenticationManager.authenticate(
+                UsernamePasswordAuthenticationToken(user.userName, user.password),
+            )
+        SecurityContextHolder.getContext().setAuthentication(authentication)
```
NIT 
more kotlinish style
```suggestion
        SecurityContextHolder.getContext().authentication = authentication
```

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/controller/AuthController.kt:
```diff
@@ -0,0 +1,69 @@
+            "IN: ${AuthController::class.java.name}: ${BASE_URL}${SIGN_IN} with userName${signInRequestDto!!.userName}",
+        )
+
+        userService.signIn(signInRequestDto.toUser())
```
That's totally insane that you need to call 2 methods from different services to authenticate user.

I suggest move security service to user service and auth user only buy `userService.signIn`

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/service/UserService.kt:
```diff
@@ -0,0 +1,56 @@
+        throw UserDoesNotExistException.fromUserName(user.userName)
+    }
+
+    fun signUp(user: User): User {
```
```suggestion
    fun signUp(user: User): User {
        try {
            return userRepository.insert(user.toMongoUser()).toUser()
        } catch (ex: DuplicateKeyException) {
            throw UserAlreadyExistException(cause = ex)
        }
    }

```

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/service/UserService.kt:
```diff
@@ -0,0 +1,56 @@
+        throw UserDoesNotExistException.fromUserName(user.userName)
+    }
+
+    fun signUp(user: User): User {
+        try {
+            return userRepository.insert(user.toMongoUser()).toUser()
+        } catch (ex: DuplicateKeyException) {
+            LOG.error(ex.toString())
+            if (ex.stackTraceToString().contains(MongoUser.USER_NAME)) {
+                throw UserAlreadyExistException.fromUserName(user.userName)
+            } else if (!ex.stackTraceToString().contains(MongoUser.EMAIL)) {
+                throw ex
+            }
+        }
+        return user
+    }
```
```suggestion
    fun signUp(user: User): User {
        try {
            return userRepository.insert(user.toMongoUser()).toUser()
        } catch (ex: DuplicateKeyException) {
            throw UserAlreadyExistException(cause = ex)
        }
    }
```

ajax-kovtunov-o commented on src/main/kotlin/io/github/anki/anki/service/UserService.kt:
```diff
@@ -0,0 +1,56 @@
+) : UserDetailsService {
+
+    @Throws(UsernameNotFoundException::class)
+    override fun loadUserByUsername(userName: String): UserDetails {
```
1) rename to getUserByUserName
2) Throws needs only if u call kotlin code from java code

sashakurta committed:
Update src/main/kotlin/io/github/anki/anki/service/UserService.kt

ну тоді це не містить функціоналу как в моем варианте а выкидывает UserAlreadyExistException всегда, а суть в том что когда мы дублируем емейл возвращать что все ок

Co-authored-by: Kovtunov Oleksandr <123635332+ajax-kovtunov-o@users.noreply.github.com>

ajax-kurta-o committed:
Fix should return 200 if user with email exist but and doesn't create one
